File: src/optical_element/MATLAB/Furion_angle_To_angle.m
function [X1,Y1,PHI,PSI] = Furion_angle_To_angle( X,Y,Phi,Psi)
% 1-project the Psi angle to yoz plane respect to z axle ;    psi-->y';
% 2-project the phi angle to xoz plane respect to z axle;     phi-->x';
%************************************************************************
% ID:210719-1 hukai hukai@mail.iasf.ac.cn
% ***********************************************************************
% Input:
% Phi(radian):
% Psi(radian):
% ***********************************************************************
% Ouput: 
% PHI(radian):
% PSI(radian):
% ***********************************************************************
% Modification history
%   2021-7-19 create hukai;
%************************************************************************
PHI = Phi;
PSI = atan(tan(Psi)./cos(Phi));
X1 = X;
Y1 = Y;
end

File: src/optical_element/Furion_angle_To_angle.cpp
#include "Furion_angle_To_angle.h"

using namespace Furion_NS;

Furion_Angle_To_Angle::Furion_Angle_To_Angle()
{

}

Furion_Angle_To_Angle::~Furion_Angle_To_Angle()
{
    
}

void Furion_Angle_To_Angle::Furion_angle_To_angle(std::vector<double>& X1, std::vector<double>& Y1, std::vector<double>& PHI, std::vector<double>& PSI, std::vector<double>& X, std::vector<double>& Y, std::vector<double>& Phi, std::vector<double>& Psi, int n)
{
    for (int i = 0; i < n; i++)
    {
        PHI[i] = Phi[i];
        PSI[i] = std::atan(std::tan(Psi[i]) / std::cos(Phi[i]));
        X1[i] = X[i];
        Y1[i] = Y[i];
    }
}

void Furion_Angle_To_Angle::Furion_angle_To_angle(std::vector<std::vector<double> >& X1, std::vector<std::vector<double> >& Y1, std::vector<std::vector<double> >& PHI, std::vector<std::vector<double> >& PSI, std::vector<std::vector<double> >& X, std::vector<std::vector<double> >& Y, std::vector<std::vector<double> >& Phi, std::vector<std::vector<double> >& Psi, int n1, int n2)
{
    for (int i = 0; i < n1; i++)
    {
        for (int j = 0; j < n2; j++)
        {
            PHI[i][j] = Phi[i][j];
            PSI[i][j] = std::atan(std::tan(Psi[i][j]) / std::cos(Phi[i][j]));
            X1[i][j] = X[i][j];
            Y1[i][j] = Y[i][j];
        }
    }
}




